### PR TITLE
Big bore air rifles

### PR DESCRIPTION
**Pull request by misterprimus**
#### Summary

```SUMMARY: Content "Adds big bore air rifles"```

#### Describe the solution

Creates "air" and associated containers and generators to facilitate it. Basically works like a UPS for the rail rifle.

#### Describe alternatives you've considered

Special checks in ranged.cpp that would jam a gun when it's out of air.
#### Additional context
N/A

### CODE REVIEW

File: data/json/items/ammo.json [hunk 1]
```diff
@@ -20,6 +20,24 @@
...
+    "price": 12000,
+    "name": "air",
+    "name_plural": "air",
+    "symbol": "=",
+    "color": "yellow",
+    "description": "Some free-floating air charge.  This can be reloaded into rechargable air cylinders, but can never be unloaded.",
+    "flags": "TRADER_AVOID",
+    "material": "iron",
+    "volume": 1,
+    "//": "Setting battery volume to 0 causes weirdness in vehicle tests.  Please don't do that.",
```
Comment (Night-Pryanik) [inline data/json/items/ammo.json]:
```suggestion
    "//": "Setting air volume to 0 causes weirdness in vehicle tests.  Please don't do that.",
```

File: data/json/items/ammo/50pellet.json [hunk 2]
```diff
@@ -0,0 +1,65 @@
+[
+  {
+    "id": "50pellet",
+    "type": "AMMO",
+    "name": ".50 Pellet",
```
Comment (Night-Pryanik) [inline data/json/items/ammo/50pellet.json]:
```suggestion
    "name": ".50 pellet",
```

File: data/json/items/ammo/50pellet.json [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    "description": "A huge .50 pellet for big bore air rifles. It causes tremendous tissue damage for an air rifle.",
```
Comment (Night-Pryanik) [inline data/json/items/ammo/50pellet.json]:
```suggestion
    "description": "A huge .50 pellet for big bore air rifles.  It causes tremendous tissue damage for an air rifle.",
```

File: data/json/items/tools.json [hunk 4]
```diff
@@ -123,6 +123,25 @@
     ],
     "magazine_well": 4
   },
+  {
+    "id": "UPS_off_air",
```
Comment (Night-Pryanik) [inline data/json/items/tools.json]: Strange id.

File: data/json/items/tools.json [hunk 5]
```diff
@@ -123,6 +123,25 @@
...
+    "type": "TOOL",
+    "name": "air fitting",
+    "name_plural": "air fittings",
+    "description": "This is an air fitting for a high pressure cylinder. It can connect a high pressure cylinder to a variety of devices utilizing air as power.",
+    "weight": 500,
+    "volume": "200 ml",
+    "price": 280000,
```
Comment (Night-Pryanik) [inline data/json/items/tools.json]: $2800 for an air fitting?

File: data/json/items/ammo/50pellet.json [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    "weight": 10,
+    "volume": 1,
```
Comment (Night-Pryanik) [inline data/json/items/ammo/50pellet.json]: 250 ml is a really high volume for a pellet.

File: data/json/items/vehicle_parts.json [hunk 7]
```diff
@@ -140,6 +140,21 @@
     "category": "veh_parts",
     "price": 90000
   },
+  {
+    "type": "GENERIC",
+    "id": "recharge_station_air",
+    "name": "air pumping station",
+    "description": "A universal recharging station designed to operate on vehicle power.  While on it will steadily fill all refillable high pressure air containers placed directly within its storage space.  The system can only be installed onto existing storage compartments, and is controlled from a dashboard or electronics control unit.",
```
Comment (Night-Pryanik) [inline data/json/items/vehicle_parts.json]:
```suggestion
    "description": "A universal air pumping station designed to operate on vehicle power.  While on it will steadily fill all refillable high pressure air containers placed directly within its storage space.  The system can only be installed onto existing storage compartments, and is controlled from a dashboard or electronics control unit.",
```

File: src/activity_handlers.cpp [hunk 8]
```diff
@@ -2076,6 +2077,25 @@ void activity_handlers::hand_crank_do_turn( player_activity *act, player *p )
 
 }
 
+void activity_handlers::hand_crank_do_turn_air( player_activity *act, player *p )
+{
+
```
Comment (Night-Pryanik) [inline src/activity_handlers.cpp]:
```suggestion
```

File: src/activity_handlers.cpp [hunk 9]
```diff
@@ -2076,6 +2077,25 @@ void activity_handlers::hand_crank_do_turn( player_activity *act, player *p )
...
+    act->moves_left -= 100;
+    item &hand_crank_item_air = p ->i_at( act->position );
```
Comment (Night-Pryanik) [inline src/activity_handlers.cpp]:
```suggestion
    item &hand_crank_item_air = p->i_at( act->position );
```

File: src/activity_handlers.cpp [hunk 10]
```diff
@@ -2076,6 +2077,25 @@ void activity_handlers::hand_crank_do_turn( player_activity *act, player *p )
...
+        p->mod_fatigue( 1 );
+        if( hand_crank_item_air.ammo_capacity() > hand_crank_item_air.ammo_remaining() ) {
+            hand_crank_item_air.ammo_set( "air", hand_crank_item_air.ammo_remaining() + 10 );
+        }
+    }
+    if( p->get_fatigue() >= DEAD_TIRED ) {
+        act->moves_left = 0;
+        add_msg( m_info, _( "You're too exhausted to keep cranking." ) );
+    }
+
```
Comment (Night-Pryanik) [inline src/activity_handlers.cpp]:
```suggestion
```

File: src/avatar_action.cpp [hunk 11]
```diff
@@ -673,6 +673,17 @@ bool avatar_action::fire_check( avatar &you, const map &m, const targeting_data
             }
         }
 
+        if( gun->get_gun_ups_drain_air() > 0 ) {
+            const int ups_drain_air = gun->get_gun_ups_drain_air();
+
```
Comment (Night-Pryanik) [inline src/avatar_action.cpp]:
```suggestion
```

File: src/avatar_action.cpp [hunk 12]
```diff
@@ -673,6 +673,17 @@ bool avatar_action::fire_check( avatar &you, const map &m, const targeting_data
...
+            if( ! you.has_charges( "UPS_off_air", ups_drain_air ) ) {
```
Comment (Night-Pryanik) [inline src/avatar_action.cpp]:
```suggestion
            if( !you.has_charges( "UPS_off_air", ups_drain_air ) ) {
```

File: src/item.cpp [hunk 13]
```diff
@@ -2594,6 +2602,10 @@ std::string item::info( std::vector<iteminfo> &info, const iteminfo_query *parts
                        parts->test( iteminfo_parts::DESCRIPTION_RECHARGE_UPSCAPABLE ) ) {
                 info.push_back( iteminfo( "DESCRIPTION",
                                           _( "* This tool has a <info>rechargeable power cell</info> and can be recharged in any <neutral>UPS-compatible recharging station</neutral>. You could charge it with <info>standard batteries</info>, but unloading it is impossible." ) ) );
+            } else if( has_flag( "RECHARGE_AIR" ) &&
+                       parts->test( iteminfo_parts::DESCRIPTION_RECHARGE_UPSCAPABLE_AIR ) ) {
+                info.push_back( iteminfo( "DESCRIPTION",
+                                          _( "* This tool has a <info>rechargeable air cell</info> and can be recharged in any <neutral>air-compatible recharging station</neutral>. You could charge it with <info>air</info>, but unloading it is impossible." ) ) );
```
Comment (Night-Pryanik) [inline src/item.cpp]: This copypaste is awful.
```suggestion
                                          _( "* This tool has an <info>air cylinder</info> and can be refilled with air in any <neutral>air pumping station</neutral>. You could fill it with <info>air</info>, but unloading it is impossible." ) ) );
```

File: src/item.cpp [hunk 14]
```diff
@@ -3402,6 +3414,10 @@ std::string item::tname( unsigned int quantity, bool with_prefix, unsigned int t
         ret << _( " (UPS)" );
     }
 
+    if( is_tool() && has_flag( "USE_UPS_AIR" ) ) {
+        ret << _( " (UPS_AIR)" );
+    }
```
Comment (Night-Pryanik) [inline src/item.cpp]: This is probably not needed at all. And even if it is, it shouldn't display internal game values, it should display user-friendly values like `(air cylinder)`.

File: src/item.cpp [hunk 15]
```diff
@@ -6430,6 +6446,9 @@ int item::units_remaining( const Character &ch, int limit ) const
     if( res < limit && has_flag( "USE_UPS" ) ) {
         res += ch.charges_of( "UPS", limit - res );
     }
+    if( res < limit && has_flag( "USE_UPS_AIR" ) ) {
+        res += ch.charges_of( "UPS_AIR", limit - res );
```
Comment (Night-Pryanik) [inline src/item.cpp]: The same here.

File: src/itype.h [hunk 16]
```diff
@@ -623,6 +625,11 @@ struct islot_battery {
     units::energy max_capacity;
 };
 
+struct islot_air {
+    /** Maximum energy the battery can store */
+    units::energy max_capacity;
```
Comment (Night-Pryanik) [inline src/itype.h]: These are not energy units. And the comment above should mention air, not the battery and not the energy.

File: src/player.cpp [hunk 17]
```diff
@@ -6894,11 +6898,25 @@ std::list<item> player::use_charges( const itype_id &what, int qty,
             res.splice( res.end(), found );
             qty -= std::min( qty, ups );
         }
+
+
+    }
+
```
Comment (Night-Pryanik) [inline src/player.cpp]:
```suggestion
```

File: src/player.cpp [hunk 18]
```diff
@@ -6894,11 +6898,25 @@ std::list<item> player::use_charges( const itype_id &what, int qty,
             res.splice( res.end(), found );
             qty -= std::min( qty, ups );
         }
+
+
```
Comment (Night-Pryanik) [inline src/player.cpp]:
```suggestion
```

File: src/player.cpp [hunk 19]
```diff
@@ -6894,11 +6898,25 @@ std::list<item> player::use_charges( const itype_id &what, int qty,
             res.splice( res.end(), found );
             qty -= std::min( qty, ups );
         }
+
```
Comment (Night-Pryanik) [inline src/player.cpp]:
```suggestion
```

File: src/player.cpp [hunk 20]
```diff
@@ -6894,11 +6898,25 @@ std::list<item> player::use_charges( const itype_id &what, int qty,
...
+
+    }
+
+    else if( what == "UPS_AIR" ) {
+
```
Comment (Night-Pryanik) [inline src/player.cpp]:
```suggestion
```

File: src/player.cpp [hunk 21]
```diff
@@ -6894,11 +6898,25 @@ std::list<item> player::use_charges( const itype_id &what, int qty,
...
+
```
Comment (Night-Pryanik) [inline src/player.cpp]:
```suggestion
```

File: data/json/vehicleparts/vehicle_parts.json [hunk 22]
```diff
@@ -2238,6 +2238,32 @@
...
+    "type": "vehicle_part",
+    "id": "recharge_station_air",
+    "name": "air pumping station",
+    "symbol": "o",
+    "color": "light_green",
+    "broken_symbol": "#",
+    "broken_color": "blue",
+    "damage_modifier": 10,
+    "durability": 20,
+    "description": "A device for refilling air cylinders.  When turned on, it charges any refillable air containers placed directly in the attached storage space.",
```
Comment (Night-Pryanik) [inline data/json/vehicleparts/vehicle_parts.json]:
```suggestion
    "description": "A device for refilling air cylinders.  When turned on, it fills any refillable air containers placed directly in the attached storage space.",
```

File: src/ammo.cpp [hunk 23]
```diff
@@ -64,7 +64,7 @@ void ammunition_type::check_consistency()
         const auto &at = ammo.second.default_ammotype_;
 
         // TODO: these ammo types should probably not have default ammo at all.
-        if( at == "UPS" || at == "components" || at == "thrown" ) {
+        if( at == "UPS" || at == "UPS_AIR" || at == "components" || at == "thrown" ) {
```
Comment (Night-Pryanik) [inline src/ammo.cpp]: `UPS_AIR` is a really bad name for this. Change it to so it doesn't mention UPS as it doesn't have any relationship to UPS at all.

**Comment by EirenexTheDragon** [comment]
wouldnt air fit well to the pneumatic weapons too? so the reload rate for putting in pellet and pumping for air could be done differently

**Comment by ampersand55** [comment]
The flamethrowers should also use a pressurised air tank as a propellant.

Currently the game flamethrowers use a pressurised fuel tank filled with flammable liquid (napalm) which is not very realistic.

* https://throwflame.com/products/x15-flamethrower/
* https://throwflame.com/products/xl18-flamethrower/

File: data/json/items/tools.json (hunk 4 above)
Comment (misterprimus) [reply data/json/items/tools.json]: The air system is strongly influenced by the UPS system and was therefore created alongside it, to make it easier to understand for others and to make future changes easier to apply to both of them. Furthermore, compressed air is used in power supplies IRL:
https://en.wikipedia.org/wiki/Compressed_air_energy_storage

though practical units were when I checked years ago prohibitively expensive.

File: data/json/items/tools.json (hunk 5 above)
Comment (misterprimus) [reply data/json/items/tools.json]: true

File: data/json/items/ammo/50pellet.json (hunk 6 above)
Comment (misterprimus) [reply data/json/items/ammo/50pellet.json]: Didn't realize those ammo files were already compatible with the new units system.

**Comment by misterprimus** [comment]
> wouldnt air fit well to the pneumatic weapons too? so the reload rate for putting in pellet and pumping for air could be done differently

You mean the one with the single round (the unmodified one)? That one has an integral pump. It doesn't come with an external inlet for an air fitting. The vast majority of the air rifles I've seen (perhaps all) have either one or the other, not both.

File: src/itype.h (hunk 16 above)
Comment (misterprimus) [reply src/itype.h]: I changed the comment, but in this case the compressed air is energy, as it is being used like stored energy, not for breathing. See compressed air energy storage.

File: data/json/items/tools.json (hunk 4 above)
Comment (Night-Pryanik) [reply data/json/items/tools.json]: > to make it easier to understand for others

It will confuse others instead as there is no relation between this system and UPS system at all.

File: data/json/items/tools.json (hunk 5 above)
Comment (Night-Pryanik) [reply data/json/items/tools.json]: That's a lot. Even if it's a RL-price, you should add a `price_postapoc` field with a much lower value as I doubt it will cost so much in a post-cataclysm world.

File: src/item.cpp (hunk 14 above)
Comment (misterprimus) [reply src/item.cpp]: I guess not. Tested the firing of a pellet with the modified gun and seems to work.

File: src/itype.h (hunk 16 above)
Comment (Night-Pryanik) [reply src/itype.h]: Don't you declare here maximum _volume_ of air cylinders, and not the compressed air energy storage value?

**Comment by EirenexTheDragon** [comment]
> 
> 
> > wouldnt air fit well to the pneumatic weapons too? so the reload rate for putting in pellet and pumping for air could be done differently
> 
> You mean the one with the single round (the unmodified one)? That one has an integral pump. It doesn't come with an external inlet for an air fitting. The vast majority of the air rifles I've seen (perhaps all) have either one or the other, not both. I view it as one of those single shot, break barrel air rifles where you break the barrel to expose the chamber for the round, load the round, and close it and it's ready to fire, so no separate air system is needed for something like that.

yes , the bolt driver and the pneumatic assault rifle! i know it has an integrated pump, but i dont know the mechanic with how its fired exactly. i thought it has a air tank thats pressurized with the pump and you just reload casually but have to repump a few times. like with Metro i played once. their description of it sounded like it fits to the pneumatic assault rifle and the bolt driver: 

"The Tihar is extremely quiet and is capable of silently bringing down enemies from surprisingly long ranges. It carries fifteen 15 mm ball bearings in a tubular spring-fed magazine. The ammunition is propelled by a cylinder of compressed air placed inside the stock of the gun, which attains its pressure by a handpump at the end of the handguard. A pressure gauge mounted on the right side of the weapon shows how much relative stopping power and distance the next shot will have, with each shot draining a small amount of air from the reservoir.

It is possible to overpressure the cylinder for extra stopping power and range, but the extra air will leak out eventually and the gun will go back to normal power levels"

so basically allow players to load in the ammunition easily like back then , but requiring the air pumping aswell which takes more time ofcourse

File: src/ammo.cpp (hunk 23 above)
Comment (misterprimus) [reply src/ammo.cpp]: Like I said it's a compressed air power supply. which is a thing. Have you read the link I posted? And it is universal in that it is fits all devices currently requiring external compressed air to function.

**Comment by misterprimus** [comment]
> The flamethrowers should also use a pressurised air tank as a propellant.
> 
> Currently the game flamethrowers use a pressurised fuel tank filled with flammable liquid (napalm) which is not very realistic.
> 
> * https://throwflame.com/products/x15-flamethrower/
> * https://throwflame.com/products/xl18-flamethrower/

Haven't used flamethrowers in a while (they seemed kind of weak), but yeah, that's definitely wrong. You can't just pour gasoline or napalm into a tank and expect it to "fire" unless you're using some rather weak Windex lever to pressurize the the contents at the time of each firing, but the range on that would be horrible. OK fir the spraycan flamethrower I guess, but anything with longer range should require a compressed air power source.

**Comment by misterprimus** [comment]
The existing pneumatic weapons may or may not need a rework (bolt driver, assault rifle). I think I have an idea of how to handle this, but the code is pretty much in the exact same section as my blackpowder PR. I don't want to create any conflicts later on, so I will wait until the blackpowder PR and this item are merged/rejected before doing any work on PCP or balancing out the existing air rifles. Likewise, flamethrower changes are likely best left to a future balancing PR.

File: src/ammo.cpp (hunk 23 above)
Comment (kevingranade) [reply src/ammo.cpp]: It is not a compressed air power supply, that's something different.
Change the name.

Comment (misterprimus) [reply src/ammo.cpp]: OK I think I whacked all the moles. Uploading now.

**Comment by misterprimus** [comment]
I'm still not sure why the monster-related "changes" are appearing on the files changed list. These are the turret ammo drop changes and they were already merged. They have nothing to do with this PR.

File: src/iuse.cpp [hunk 24]
```diff
@@ -4255,6 +4255,39 @@ int iuse::hand_crank( player *p, item *it, bool, const tripoint & )
...
+    }
+    if( p->get_fatigue() >= DEAD_TIRED ) {
+        p->add_msg_if_player( m_info, _( "You're too exhausted to keep cranking." ) );
+        return 0;
+    }
+    item *magazine = it->magazine_current();
+    if( magazine && magazine->has_flag( "RECHARGE_AIR" ) ) {
+        int time = to_moves<int>( 1600_minutes );
+        if( it->ammo_capacity() > it->ammo_remaining() ) {
+            p->add_msg_if_player( string_format( _( "You start cranking the %s to charge its %s." ),
```
Comment (BevapDin) [inline src/iuse.cpp]: `add_msg_if_player` does it's own string formatting, you don't need to do it yourself.

File: src/ranged.cpp [hunk 25]
```diff
@@ -326,6 +326,11 @@ int player::fire_gun( const tripoint &target, int shots, item &gun )
         shots = std::min( shots, static_cast<int>( charges_of( "UPS" ) / gun.get_gun_ups_drain() ) );
     }
 
+    if( !gun.has_flag( "VEHICLE" ) && gun.get_gun_compressed_air_drain() > 0 ) {
+        shots = std::min( shots, static_cast<int>( charges_of( "compressed_air" ) /
```
Comment (BevapDin) [inline src/ranged.cpp]: The static_cast shouldn't be necessary.

File: src/visitable.cpp [hunk 26]
```diff
@@ -812,6 +816,11 @@ int visitable<inventory>::charges_of( const std::string &what, int limit,
         qty = sum_no_wrap( qty, static_cast<int>( charges_of( "adv_UPS_off" ) / 0.6 ) );
         return std::min( qty, limit );
     }
+    if( what == "compressed_air" ) {
+        int qty = 0;
+        qty = sum_no_wrap( qty, charges_of( "compressed_air_container" ) );
```
Comment (BevapDin) [inline src/visitable.cpp]:
```suggestion
        int qty = charges_of( "compressed_air_container" );
```

File: src/visitable.cpp [hunk 27]
```diff
@@ -855,6 +864,12 @@ int visitable<Character>::charges_of( const std::string &what, int limit,
         return std::min( qty, limit );
     }
 
+    if( what == "compressed_air" ) {
+        int qty = 0;
+        qty = sum_no_wrap( qty, charges_of( "compressed_air_container" ) );
```
Comment (BevapDin) [inline src/visitable.cpp]: Same here.

**Comment by Rivet-the-Zombie** [comment]
Some kind of gunmod that would convert an extant pneumatic weapon to require the separate air canister in exchange for drastically reduced reload time would be capital.

**Comment by misterprimus** [comment]
> Some kind of gunmod that would convert an extant pneumatic weapon to require the separate air canister in exchange for drastically reduced reload time would be capital.

The existing air guns need to be totally redone IMO. I am using the pneumatic assault rifle and with bearings it's actually doing more damage than a 9mm gunpowder handgun! IRL it's 9mm > .22 > .177 air rifle. I think it's a bit outside the scope of this PR, though. If it were up to me I'd simply nerf that pneumatic assault rifle to little more than a kid's toy. Any air rifle that rises to the level of a weapon is either single shot or precharged. Anything else and you just don't get any stopping power, which is questionable with anything but a top shelf single shot break barrel.

**Comment by Rivet-the-Zombie** [comment]
>The player would be considered passively pumping the pneumatic as he fires since it would be weaker.

That's why they have such long per-shot reload times - to account for the time spent pumping the gun's tank up a bit more each time.

>Oh and the nail rifle really has to go away.

I'm not a fan of nuking any of our fun scrap-built weapons. Making the nail rifle require an air tank would make sense, though.

>get rid of everything else pneumatic, incl. flamethrowers and the bolt driver.

Oh heck no.

**Comment by misterprimus** [comment]
> > The player would be considered passively pumping the pneumatic as he fires since it would be weaker.
> 
> That's why they have such long per-shot reload times - to account for the time spent pumping the gun's tank up a bit more each time.
> 
> > Oh and the nail rifle really has to go away.
> 
> I'm not a fan of nuking any of our fun scrap-built weapons. Making the nail rifle require an air tank would make sense, though.
> 
> > get rid of everything else pneumatic, incl. flamethrowers and the bolt driver.
> 
> Oh heck no.

I don't mean permanently, I just mean until a realistic option can be made, like for the flamethrower. Nail guns need to go away because nails themselves make bad projectiles. There's no way you can just pick up a bunch of "nails," load them into a nail rifle, and shoot 10 in an instant/burst, effectively killing whatever is in front of you. At best you'll get ineffective semi-auto operation unless the gun is truly gas powered, but that would only work with gunpowder-powered nail guns. And even those don't do much damage at any range.

**Comment by ymber** [comment]
This is the same kind of hack that UPS pseudo ammo currently works by. Properly handling pneumatic items is addressed in the multimag project. Look there for how it needs to be implemented. As it is this doesn't bring us closer to the established goals for ammunition handling and will make future work harder.

**Comment by misterprimus** [comment]
> This is the same kind of hack that UPS pseudo ammo currently works by. Properly handling pneumatic items is addressed in the multimag project. Look there for how it needs to be implemented. As it is this doesn't bring us closer to the established goals for ammunition handling and will make future work harder.

Correct it's based on how the UPS works for the ferromagnetic rifle.

File: data/json/items/generic.json [hunk 28]
```diff
@@ -2929,6 +2929,67 @@
     "weight": 11339,
     "volume": 12
   },
+  {
```
Comment (John-Candlebury) [inline data/json/items/generic.json]: I only gave your code a quick glance, but the changes in this file seem completely unrelated to adding the big bore rifles?

If they are unrelated, you should probably revert them and split them into a new pr.

Comment (misterprimus) [reply data/json/items/generic.json]: They're changes from my turret ammo drops PR. I don't know how they got into this PR or even why they appear as changes since that PR was already merged. Do I just delete those files from my PR?

**Comment by ymber** [comment]
The way the UPS works now is a pervasive problem all over the code and copying how it works is not an acceptable way to implement any new features. This is a large change set that we're going to have to rip out again when we get to the relevant part of multimag because the way this is designed is fundamentally incompatible with the direction we're going in for ammunition handling. I'm closing this because it's too flawed to be desirable in a recognizable form.

See #32227.

**Comment by kevingranade** [comment]
I apologize for not reviewing this before, but on reading @ymber 's rationale, I agree with @ymber that we really do not want to follow this approach, we need to treat compressed air as a separate ammo type within the gun itself, not in an external tank.

File: data/json/items/generic.json (hunk 28 above)
Comment (ymber) [reply data/json/items/generic.json]: For future reference the way to deal with this is update your master and rebase your feature branch onto it.

**Comment by misterprimus** [comment]
> I apologize for not reviewing this before, but on reading @ymber 's rationale, I agree with @ymber that we really do not want to follow this approach, we need to treat compressed air as a separate ammo type within the gun itself, not in an external tank.

I suppose I thought it was a decent solution since it works quite well. As it is air guns range from badly unbalanced (pneumatic bolt driver, pneumatic assault rifle) to downright impossible (nail rifle on burst) and I thought this would be a good way to allow those to either be made more realistic or removed entirely to be replaced with the two guns I included in this PR, particularly since it would require little extra work to rip this stuff out (it was intentionally placed next to the UPS energy use code, which will have to be ripped out anyway I assume).

Hopefully the multimag PR is finished soon.

**Comment by misterprimus** [comment]
> I apologize for not reviewing this before, but on reading @ymber 's rationale, I agree with @ymber that **we really do not want to follow this approach, we need to treat compressed air as a separate ammo type within the gun itself**, not in an external tank.

Also, just wanted to follow up, I was planning on making another rifle that is a PCP multishot that does have a separate parameter for internal air, capable of being recharged by activating the air fitting item (and using up the charges in the container). The present modified rifle in this PR is basically this:
https://www.youtube.com/watch?v=BeeJ6OQzTAc
without an integral air storage container.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/pressure-for-pneumatic-weapons/21289/13
